qa/tasks/cephfs/test_volumes.py

```diff
@@ -676,6 +676,102 @@ def test_volume_create(self):
         self._fs_cmd("volume", "rm", volname, "--yes-i-really-mean-it")
 
+    def test_with_both_pool_names(self):
+        '''
+        Test that "ceph fs volume create" command accepts metadata pool name
+        and data pool name as arguments and uses these pools to create a new
+        volume.
+        '''
+        v = self._gen_vol_name()
+        meta = 'meta4521'
+        data = 'data4521'
+        self.run_ceph_cmd(f'osd pool create {meta}')
+        self.run_ceph_cmd(f'osd pool create {data}')
+        self.run_ceph_cmd(f'fs volume create {v} --data-pool {data} '
+                          f'--meta-pool {data}')
+        o = self.get_ceph_cmd_stdout('fs ls')
+        o = o.split('\n')
+        for i in o:
+            if f'name: {v},' in o:
+                self.assertIn(f'metadata pool: {meta}')
+                self.assertIn(f'datadata pool: [{data} ]')
+
+    def test_with_data_pool_name_only(self):
+        '''
+        Test that "ceph fs volume create" command runs successfully when data
+        pool name is passed, the data pool name aborts with an complain about
+        not passing metadata pool name.
+         '''
+        v = self._gen_vol_name()
+        data = 'data4521'
+        self.run_ceph_cmd(f'osd pool create {data}')
+
+        self.negtest_ceph_cmd(f'fs volume create {v} --data-pool {data}',
+                              retval=errno.EINVAL,
+                              errmsgs=('metadata pool name isn\'t passed'))
+
+        o = self.get_ceph_cmd_stdout('fs ls')
+        self.assertNotIn(f'name: {v},', o)
+
+    def test_with_metadata_pool_name_only(self):
+        '''
+        Test that when only metadata pool name is passed to "ceph fs volume
+        create" command, the command aborts with an error complaining about
+        not passing data pool name.
+        '''
+        v = self._gen_vol_name()
+        meta = 'meta4521'
+        self.run_ceph_cmd(f'osd pool create {meta}')
+
+        self.negtest_ceph_cmd(f'fs volume create {v} --meta-pool {meta}',
+                              retval=errno.EINVAL,
+                              errmsgs=('data pool name isn\'t passed'))
+
+        o = self.get_ceph_cmd_stdout('fs ls')
+        self.assertNotIn(f'name: {v},', o)
+
+    def test_with_nonempty_meta_pool_name(self):
+        '''
+        Test that when meta pool name passed to the command "ceph fs volume
+        create" is an non-empty of pool, the command aborts with an appropriate
+        error number and error message.
+        '''
+        v = self._gen_vol_name()
+        meta = f'cephfs.{v}.meta'
+        data = f'cephfs.{v}.data'
+
+        self.run_ceph_cmd(f'osd pool create {meta}')
+        self.run_ceph_cmd(f'osd pool create {data}')
+        self.mon_manager.controller.run(args='echo somedata > file1')
+        self.mon_manager.do_rados(['put', 'obj1', 'file1', '--pool', meta])
+        # XXX
+        log.info('sleeping for 10 secs for stats to be generated so that "fs '
```

Comment:
Which "stats" are you referring to?

Reply:
`mon->mgrstatmon()->get_pool_stat(metadata)` specifically.

Reply:
So, technically, if a user creates a pool and dumps some objects in it, the non-empty pool can still be used to create the file system since there is a lag in the reported stats? Is there a way to explicitly refresh the stats before creating the file system?

Reply:
> the non-empty pool can still be used to create the file system since there is a lag in the reported stats? 

For a very small time (like a second or two), yes.

> Is there a way to explicitly refresh the stats before creating the file system?

I went through the code that fetch stats and didn't find an option to refresh stats. I don't think it's about refreshing too, it's about the fact generating stats for OSD pool takes some time, just like MDS rstats.

Reply:
So, it's basically possible to sneak in a volume create (with existing pools) in that time windows. Practically, it's probably not an issue, but that depends on the time window itself, which could be longer, maybe - so, I'm worries _a bit_.

Reply:
As you said, this check is already in MDSMonitor.cc, so its not directly an issue with this change, but is there something better than we can do to avoid this window?

Reply:
> So, it's basically possible to sneak in a volume create (with existing pools) in that time windows. Practically, it's probably not an issue, but that depends on the time window itself, which could be longer, maybe - so, I'm worries _a bit_.

I agree. I was about to propose a change in the MON code here. It would be better to not depend on stats and here instead do list objects in pool. If there are any, we can error out saying "poo non-empty". Unlike stats, this method suffers from no lag.

Reply:
You could have a (slow path) validation to check for at least one object in the pool if the earlier (fast path) check using stats returned no object count.